### DISCUSSION TITLE
4k youtube playback, hitches, stutter, framedrops

### DISCUSSION THREAD

**Discussion by geextahslex** [mpv-player/mpv#13991] [Q&A]
### Important Information

Provide following Information:
- mpv v0.38.0
- yt-dlp 2024.04.09
- Windows 10 22H2 (Build 19045.4170)
- shinchiro from Sourceforge.net
- I started using mpv 36 so since then I know this issue
- intel HD 4600, driver and version 20.19.15.5171 (15.40.5171)
- screenshot
![mpvdssd](https://github.com/mpv-player/mpv/assets/139760266/f15acf45-353e-40e9-bc80-d66257f474c4)
![mpvdssd 2](https://github.com/mpv-player/mpv/assets/139760266/08ad1b04-ce40-43ec-a7b2-b23d6d7f9fcd)

### Reproduction steps

Start playback of a 4k 60 fps youtube video with latest yt-dlp

### Expected behavior
Normal playback, stutterfree

### Actual behavior
Random hitches, stutter while playback

### Log file

[output.txt](https://github.com/mpv-player/mpv/files/15111709/output.txt)

### Sample files
used youtube video
https://www.youtube.com/watch?v=xmJUZgQ5HpQ

### Side note
the behaviour is the same when I downloaded the same file and watched it from the HDD
So yt-dlp/webstream can be excluded as a source of the stutters

**Comment by kasper93** [comment]
`--hwdec=yes`

**Comment by YukiNeeSan** [comment]
[   0.008][v][cplayer] Applying profile 'default'...
[   0.008][v][cplayer] Setting option 'profile' = 'fast' (flags = 4)
[   0.008][v][cplayer] Applying profile 'fast'...

`--hwdec=auto-safe or --hwdec=auto-copy` sometimes copy-back works well on Intel i-GPU based while auto-safe fit with AMD APU based Processor,
don't forget put `--vo=gpu-next or --vo=gpu`

**Comment by geextahslex** [comment]
>`hwdec=yes`

doesn't make a difference

>`vo=gpu-next`

I already am using this

**Comment by llyyr** [comment]
`--profile=fast`

**Reply by kasper93** [reply]
> `--profile=fast`

Their GPU is at 15-20% I doubt it is rendering bottleneck. 

@geextahslex Could you show stats page 2? shift+i and 2

Also does it happen with local files? Maybe you see buffering issue, those can happen with youtube. Check page 3 of stats to see if you have healthy cache.

**Reply by geextahslex** [reply]
>Also does it happen with local files?

Yes I have stated that in the last part of my post
>Side note
the behaviour is the same when I downloaded the same file and watched it from the HDD
So yt-dlp/webstream can be excluded as a source of the stutters

Here is the screenshot, stats page 2
![stats2](https://github.com/mpv-player/mpv/assets/139760266/67187503-4b98-4fa3-8213-a98f7e4814c9)


@llyyr 
I am already using `profile=fast
`

**Reply by kasper93** [reply]
Looks like decoding performance issue, can you make sure `hwdec` is used? Page 1 or log to confirm.

**Reply by geextahslex** [reply]
Where do you see if hwdec is on or off?
Also I don't understand why dropped frames is at 0 when there are these hitches etc.
![hwdec](https://github.com/mpv-player/mpv/assets/139760266/43a91b38-b6fe-4513-ac2e-450a50b1f565)
![cuda](https://github.com/mpv-player/mpv/assets/139760266/338d2d26-ef1a-4b91-8420-1fb8238784bf)
[output.txt](https://github.com/mpv-player/mpv/files/15113334/output.txt)

**Reply by YukiNeeSan** [reply]
hardware decoding enable it should be there... 
![Screenshot 2024-04-25 103433](https://github.com/mpv-player/mpv/assets/79769464/033bb9e1-271e-41e0-b141-bc1108cd327d)
![Screenshot 2024-04-25 103801](https://github.com/mpv-player/mpv/assets/79769464/8c5d000e-3627-4729-aa9d-6a807265fdac)


Hmmm when tried on the weaker system get a lot of frame drops even there's not eat system resource too much downloaded [the vp9 google video](https://youtu.be/Mw1MWfilz_k?si=26NJobcOUlqKu8TS) 
![Screenshot 2024-04-25 104541](https://github.com/mpv-player/mpv/assets/79769464/eea463cc-ecc7-441a-a2c1-007249347299)

[FSR.zip](https://github.com/mpv-player/mpv/files/15119120/FSR.zip)
put in FSR (EASU) on the config there's no frame drops or hiccups, out of topic but the usage of system resource still no different, strange

**Comment by na-na-hi** [comment]
> * intel HD 4600

Doesn't support VP9 hwdec. Your hardware is probably too weak for 4K60 YouTube playback.

**Comment by geextahslex** [comment]
I don't know if it is to weak, maybe. But when I limit the FPS to 30, the CPU is at 50-60% and the GPU at 15% load, there is still jitter all over the place. I can watch 4k HDR with tonemapping with no issues.
I think it should be either capable of running it or not. I watched many 4k60 videos and it runs smoothly for 95% of the time, but now and then there are some hitches. I feel like there is a optimazation issue or something. Neither CPU nor GPU are hitting 100%
![limited](https://github.com/mpv-player/mpv/assets/139760266/2a6f2a10-8874-4655-ba59-6c0168426a9c)

### Bootleneck representation with slapped filters on it (HDR tonemapping, debanding, sharpening, denoise, glsl saturation, vibrance shaders)

### CPU bottleneck
When CPU limited, there are small variations (delta between max and min frame) but overall the FPS is at a constant level, without random spikes 
![cpu bottleneck](https://github.com/mpv-player/mpv/assets/139760266/d6714bb5-5a60-4216-9fb5-37e77b8876c4)


### GPU bottleneck
When GPU limited you have an almost flat FPS line at the performance you can hit, no jitter
![gpu bottleneck](https://github.com/mpv-player/mpv/assets/139760266/0ceec29c-f33e-4bb0-8967-7c9ade5e176a)

### 4k60FPS
flat 60
![4k60 ff gg](https://github.com/mpv-player/mpv/assets/139760266/ca2597e1-508d-4ca0-a792-4dd029fc6d5f)
60FPS is fine but it is interrupted by these random spikes

![4k60 ff](https://github.com/mpv-player/mpv/assets/139760266/0db486b7-6bf3-4366-a492-d1df1c9812a8)

This is reproducible, everytime at this scene (02:22) the playback literally stops ^^
![framdrop](https://github.com/mpv-player/mpv/assets/139760266/0915437b-c66f-49ab-9882-27140f25c0e9)

**Reply by na-na-hi** [reply]
>This is reproducible, everytime at this scene (02:22) the playback literally stops ^^

The amount of CPU needed for decoding depends on the scene complexity. For highly dynamic scenes more CPU is needed. It seems that your CPU is barely powerful enough that it can handle most scenes, but not for these specific ones. That's how these reproducible "jitters" happen.

**Reply by kasper93** [reply]
You could easily plot the cpu usage and see if there is corelation visible with "jitters". I would expect so. Also if your hardware is not capable of VP9 hardware decoding, set ytdl to prefer h264, you will have a lot better experience. Lastly, check other players like VLC or MPC-HC if they behave better.

**Reply by geextahslex** [reply]
@kasper93 I monitored each core (4) usage while playback and no core hits above 87%. 
Isn't it possible to make a logfile for this specific scene 10-20 seconds, and actually see what is causing this? Or does mpv give not enough data feedback (on hardware/performance/monitoring level) to see whats going on?

**Reply by kasper93** [reply]
87% is very high for media playback and if it is sporadic higher load you may not see it in usage% which is often smoothed.

> Isn't it possible to make a logfile for this specific scene 10-20 seconds, and actually see what is causing this? Or does mpv give not enough data feedback to see whats going on?

It is possible to debug it with profiler (Visual Studio profiler would be easiest on Windows) and seem where is the bottleneck, but it is not something that is done by mpv itself, at least not with modification.

**Reply by geextahslex** [reply]
Here is a scene with a still image (opened map) and you can see the cpu is at 30-50% usage. But you can count 6 jitters, 2 you can see on the graph in the background. There seems to be a pattern to those dopped frames.
Also this doesn't make sense that a still image has more jitter than normal gameplay with running around/movement.
![fsfef](https://github.com/mpv-player/mpv/assets/139760266/2da0e3dc-596f-4a21-9668-6e26664b203f)

**Comment by geextahslex** [comment]
Okay and why is this video running smoothly? It's 4k60 VP9
![stellar blade](https://github.com/mpv-player/mpv/assets/139760266/0bf41be8-8dae-43b6-94a8-75114f0e11a3)
https://www.youtube.com/watch?v=rItLyefVK3A

**Comment by geextahslex** [comment]
I have noticed eacht time there is this pattern the framerate drops from 60 to 8, isn't that a bit drastic?
This would mean that depending on the scene the hardware needs at least 650% more performance.

![pattern](https://github.com/mpv-player/mpv/assets/139760266/152c0b39-ffcd-45e9-b19c-52df6098b124)
zoomed in
![pattern2](https://github.com/mpv-player/mpv/assets/139760266/63d1f892-b181-40d2-9a9f-06b0968f3b27)
source: https://www.youtube.com/watch?v=O-x_d2V9PDY 
Time 49:32

**Reply by na-na-hi** [reply]
It seems that you have a dual-core CPU model (shown as 4 CPUs due to hyperthreading). In this case it's simply underpowered, and hyperthreading doesn't help much so when it reaches "50%" usage it's really close to 100%. Your CPU usage shows that it's fully utilized. 

And yes this type of scene is very CPU-intensive.

**Reply by geextahslex** [reply]
4590 is 4 core, there is no HT

**Reply by na-na-hi** [reply]
Well this sample drops frames even with only 12% CPU usage on my setup, I can only play it at 0.9x speed without frame drops. No problem with hardware decoding. 

I guess it's yet another case of FFmpeg being shit.

**Reply by geextahslex** [reply]
Okay thats interesting. Yeah this seems to me like a "bug" or optimization issue.

**Reply by kasper93** [reply]
You could try AV1 version if VP9 have some weirdness.

> I guess it's yet another case of FFmpeg being shit.

Would be interesting to see where exactly does it choke.

**Reply by geextahslex** [reply]
@kasper93 The default yt-dlp setting already sets AV01 above VP9
vcodec: Video Codec (av01 > vp9.2 > vp9 > h265 > h264 > vp8 > h263 > theora > other)

**Reply by hrxn** [reply]
Default yt-dlp picks AV1 above VP9? That would be news to me.

Are you sure it is not your config?

Also, what is the 'problematic' sample?
This one? https://www.youtube.com/watch?v=O-x_d2V9PDY

**Reply by geextahslex** [reply]
@Hrxn I don't know, this information is on their github page
yes, stellar blade drops from 60 to 8 fps when there is this white scanning thing
and the other sample https://www.youtube.com/watch?v=xmJUZgQ5HpQ has hitches all the time
